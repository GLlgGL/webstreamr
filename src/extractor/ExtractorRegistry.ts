import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, CountryCode, UrlResult } from '../types';
import { isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult[]>;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;
    this.urlResultCache = new TTLCache({ max: 4096 });
  }

  public async handle(ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    let urlResults = this.urlResultCache.get(cacheKey) ?? [];
    if (this.urlResultCache.has(cacheKey)) {
      return urlResults.map(urlResult => ({ ...urlResult, ttl: this.urlResultCache.getRemainingTTL(cacheKey) }));
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    urlResults = await extractor.extract(ctx, normalizedUrl, countryCode, title);

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      this.urlResultCache.set(cacheKey, urlResults, { ttl: extractor.ttl });
    }

    return urlResults;
  };
}
