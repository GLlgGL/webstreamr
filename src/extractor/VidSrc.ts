import * as cheerio from 'cheerio';
import slugify from 'slugify';
import { BlockedError, NotFoundError, TooManyRequestsError } from '../error';
import { Context, Format, Meta, NonEmptyArray, UrlResult } from '../types';
import { Fetcher, guessHeightFromPlaylist } from '../utils';
import { Extractor } from './Extractor';

export class VidSrc extends Extractor {
  public readonly id = 'vidsrc';

  public readonly label = 'VidSrc';

  public override readonly ttl: number = 10800000; // 3h

  private readonly tlds: NonEmptyArray<string>;

  public constructor(fetcher: Fetcher, tlds: NonEmptyArray<string>) {
    super(fetcher);

    this.tlds = tlds;
  }

  public supports(_ctx: Context, url: URL): boolean {
    return null !== url.host.match(/vidsrc/);
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    return this.extractUsingRandomTld(ctx, url, meta, [...this.tlds]);
  };

  private async extractUsingRandomTld(ctx: Context, url: URL, meta: Meta, tlds: string[]): Promise<UrlResult[]> {
    const tldIndex = Math.floor(Math.random() * tlds.length);
    const [tld] = tlds.splice(tldIndex, 1) as [string];

    const newUrl = new URL(url);
    const hostnameParts = newUrl.hostname.split('.');
    hostnameParts[hostnameParts.length - 1] = tld;
    newUrl.hostname = hostnameParts.join('.');

    let html: string;
    try {
      html = await this.fetcher.text(ctx, newUrl, { headers: { Referer: meta.referer ?? newUrl.origin }, queueLimit: 1 });
    } catch (error) {
      if (tlds.length && (error instanceof TooManyRequestsError || error instanceof BlockedError)) {
        return this.extractUsingRandomTld(ctx, url, meta, tlds);
      }

      throw error;
    }

    const $ = cheerio.load(html);

    const iframeUrl = new URL(($('#player_iframe').attr('src') as string).replace(/^\/\//, 'https://'));
    const headers = { Referer: iframeUrl.origin };
    const title = $('title').text().trim();

    return Promise.all(
      $('.server')
        .map((_i, el) => ({ serverName: $(el).text(), dataHash: $(el).data('hash') }))
        .toArray()
        .filter(({ serverName }) => serverName === 'CloudStream Pro')
        .map(async ({ serverName, dataHash }) => {
          const iframeHtml = await this.fetcher.text(ctx, new URL(`/rcp/${dataHash}`, iframeUrl.origin), { headers });
          const srcMatch = iframeHtml.match(`src:\\s?'(.*)'`) as string[];

          const playerHtml = await this.fetcher.text(ctx, new URL(srcMatch[1] as string, iframeUrl.origin), { headers });
          const fileMatch = playerHtml.match(`file:\\s?'(.*)'`);
          if (!fileMatch) {
            throw new NotFoundError();
          }

          const m3u8Url = new URL(fileMatch[1] as string);

          return {
            url: m3u8Url,
            format: Format.hls,
            label: `${this.label} (${serverName})`,
            sourceId: `${this.id}_${slugify(serverName)}_${meta.countryCodes?.join('_')}`,
            ttl: this.ttl,
            meta: {
              ...meta,
              height: await guessHeightFromPlaylist(ctx, this.fetcher, m3u8Url),
              title,
            },
          };
        }),
    );
  }
}
