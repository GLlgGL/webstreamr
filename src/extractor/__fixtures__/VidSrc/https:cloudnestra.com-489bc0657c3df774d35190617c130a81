<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html"/>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.0-2/css/all.min.css" integrity="" crossorigin="anonymous"/>
<link rel="stylesheet" href="/style_rcp-e600e6.css?t=1744906951"/>

<style>
    body {
        font-family: system-ui, -apple-system, "Segoe UI", Roboto, "Helvetica Neue", "Noto Sans", "Liberation Sans", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
    }
    .hidden {
        display:none;
    }
    
    #asdf {
        width: 0px;
        height: 0px;
    }
    
    #iframe_title {
        position: absolute;
        top: calc(50% + 45px);
        font-size: 1.1em;
        text-shadow: -1px -1px 0 #000, 1px -1px 0 #000, -1px 1px 0 #000, 1px 1px 0 #000;
        width: calc(100% - 30px);
        text-align: center;
        padding: 0px 15px;
    }
</style>
</head>
<body data-i="93058"  >
    
    
        <!--<div id="pop_asdf" style="position:absolute; width:100%; height:100%; z-index:2147483650; pointer-events: none;"></div>-->
        
    <!--<div id="loading_overlay"></div>-->
    <div id="the_frame" style="background-image: none;">
                <div id="pl_but_background" style="background-image: url('//image.tmdb.org/t/p/w780/yYderAvjtKnYLBxKujVpmKTVlOM.jpg');">
            <div id="iframe_title">
                Full Metal Jacket (1987)            </div>
            <i id="pl_but" class="fas fa-play"></i>
        </div>
            </div>
    
    <div id="hidden" style="display:hidden;"></div>
    
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js" integrity="sha256-1A78rJEdiWTzco6qdn3igTBv9VupN3Q1ozZNTR4WE/Y=" crossorigin="anonymous"></script>
    <script src="/base64.js?t="></script>
    
    <script src="/sbx.js?t=1751380596"></script>

    
    <script>
    
    /*
    $(document).ready(function(){
        
        
        if($.cookie('pop_asdf')){
            $("#pop_asdf").addClass("hidden");
        }
        
        $("#pop_asdf").click(function(){
            if(!$.cookie('pop_asdf')){
                if($.cookie('pop_asdf_tmp') >= 3){
                    var date = new Date();
                    date.setTime(date.getTime() + (10800 * 1000));
                    $.cookie('pop_asdf', 1, { expires: date });
                    $("#pop_asdf").addClass("hidden");
                }else{
                    var cookie_value = 1;
                    if($.cookie('pop_asdf_tmp'))
                        cookie_value = $.cookie('pop_asdf_tmp');
                        
                    cookie_value++;
                    
                    $.cookie('pop_asdf_tmp', cookie_value);
                    $(this).addClass("hidden");
                    setTimeout(function(){
                        $("#pop_asdf").removeClass("hidden");
                    }, 59000);
                }
            }
        });
    });
    */
    
        
    $("#pl_but_background , #pl_but").click(function(){
        loadIframe();
    });
    
        
    function loadIframe(data = 1){
        if(data == 1){
            $("#the_frame").removeAttr("style");
            $("#the_frame").html("");
            $('<iframe>', {
               id: 'player_iframe',
               src: '/prorcp/YjBhNDIwZjYxYzcwZDk4YWI0OWJhZWVhNzU1ODBlZDk6YUVGNmFuVkxlRXNyZW0xNGJXTTRVV2hCY0docFZrUlFObXBXWjJrd2MzUlBUbXBsYlVvemFXUjVXSGRpTVRaaFdVMTJkR2hXUmxwdFpYZ3JiVXh1Y0ZvMWNrOVJMMjlVVlhaeU5FUkRPVVl3Y0hkMFRtdG5VRFJRVkhwWGRVaHlaRTFWVTNVMlRsVTRhakZHZDNGVmEzcHdRbHBCUmxSbVozcFBWbVpLVTA5SFpYZDFPVWR1TUdSMGQyeGhWMnRoY1RSSlZXSXlaRWhQYnpJelEzTlpZVloxU0M5eU1FaEJWMlZLZHpoc1RWWXdVelJVUm1sUVJFOVpaMmRIUkhnNGFVOWlVMGhwUTFad1JEVTBLM0pvVVhweVRrSlNSRXh0YW1SRFpXNXVOWE13VFRkVFJHNTBXRkpPYjB0WE1HTnNlVXA0TkdKVE0zbEhXR3M0TTFCTFVUbExSemxIZFVwSFdIaHhjREpXUldkUFdsRXdaVU42UmtKclkwcG5WVE40Ym5WU1RrTlhhVFpUU0c5WlNUQXdXamM1Y21GWFVXVTJZM0JHTkVzNWVUTmpiVll2UVRGa1ZGQkZlVmhpYkZOaWFrMW9ZVFpoYkZZM1dXUXdjREp3TUdWUmNEZFFTbGdyY25oNk5FSjZVVVZwUkhneU5pOXhjMkpvU1VGVk0wRkZPRXR3YVZBdk5GbFNOSE5NWkhKRmRFNWhlazk0UlZKVGJGZE1VVVk0TmxaVlpVcEVaR042WWxKU1ZqQXJNbEpLVDJsQlFrbGxPR3hZVmpabGNVSTFibVJuVUVZeU5XOW5iRmgxTVd0Vk1rdFdORXhQT1RKcFRVNHdZMHd3VDFSeFZGQkRXa0ZOTTFsblVHMHpiMmxXYWxCWlNHSXZNQ3MyTDNSYVJqaG9ZM0J1VmtaYWQzQXpOR1Z0VGpOelEyOU5XWFZtY0ZkcVJuaFFjRFpqTURVMVMyaENTbWh3U2pjM1ZucFNOWFJtV1RWMmMybHJZVUV3VG5kdFVtNXlha2x2YmxFd2NqQndSemwxWVRKS05EUjFjVUpoZEM5M2MzUklTRXg1VEZaVGJsaFpNa3RJYkdaaFQyeDRORkZNVEV0dVNuVTNRMU5oZUdoNFJVTmxhRzVyZDNKYU5uVTFXbm8yUlRKd1YxbDBiMkY0TkhoeFYwSnBhMHhZY0ZOU1J6RjVWWEJRYjNGeWVsTmFVV2hqYmtNNGVYTkpjMW80WjFsT1RrUTJNVlJNTjNkSGRYSnhNM1l3YjNOUlVDdDJVRmd2WmxWRU9HTmtWVTV6WmpNclVXcFVWV1J2WjJFdmQycDZlVkZKUkdsS2NVVnlZM2hEVW0xQlkwbGtOekkxZW5WalZETTRTMVUzTmpsUVNHd3ZaVXBuU0ZWSWVWVnVObVl6Tm1SRVUyUTFRV0Z0WkRkWWNIWkNkSFEyVUdsMFUweDVWM05qZFRKTVVFSmpVbkZTYm5odk5sRlZTRXhpZHpKUWNrNXdSbk5EUjJOM2JsTktZbVJoUjJOV1REbG9lbVFyS3preEwyVmpkMGQxWjNSdWVIUlNOME0zYzBseloycFRjekZYYWxkNVowRkVjVGRRTmtjeFFtTjNVRVF3T0hWdlowMW9lUzkyTlVSbFZFdDBlRTh5VWs1dlNEazROeXM0UWpGS2JuRTJaSFZ5UWpoMlRTdHpVV2N3YzFGTmNXRldVV05xVGtsb1JIbE9Oa2cwYjNSb1FYUnNVVXh4YjBaU0t6Qm1iQ3RFVm0xVllrZERNMEpsVEZGamVHbEJjMGhrWTNSMmQzSnFOMWszY1VKV1prWmphMmhNVVRkVVRsWjZTbFp2VG10T2IyVXhkblZJVUU4MU5XUTRPV3R1V0dzd1luRkNOSFJNV2xKSE5saG5jMDk1Tm1sTVVYSnlVWGhvVkVwWU1taGhUMFZXWlVWM1F6VkxkalJUTUdGRk0xVmlRWFJqUlZselltNUhOWE5KTlZBeFEzcEdWalU0VUROU01sZFpWblV3ZDFsd1RERm1RMjlYV0U5T2RXTlRlbEZWY1V4TUx5dHBZbmhzZUZNeFkxcHBaa1ZpZW1GaFkzb3djbEp5UzJkUGNIaHlWMDB6V0dZd2FubHhWbTl2TkdSNFRrcG9jRTkzYzFOT1JURnNja015TWc9PQ--',
               frameborder: 0,
               scrolling: 'no',
               allowfullscreen: 'yes',
               allow: "autoplay",
               style: 'height: 100%; width: 100%;'
            }).appendTo('#the_frame');
            $("#player_iframe").on("load", function () {
                $("#the_frame").attr("style","background-image: none;");
            });
        }
    }
    
    // pm redirector
    window.addEventListener('message', message => {
        if (message.source == window) {
            return; // Skip message in this event listener
        }
        
        var the_iframe = document.getElementById('player_iframe');
        
        if(message.source == window.parent){
            the_iframe.contentWindow.postMessage(message.data,'*');
        }else{
            window.parent.postMessage(message.data , '*');
        }
        
        
        
    });
        
        
        
    </script>
    
    
    
    
    
    
   
</body>
</html>
