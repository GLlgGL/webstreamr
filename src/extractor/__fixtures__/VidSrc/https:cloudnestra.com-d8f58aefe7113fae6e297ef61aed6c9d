<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html"/>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.0-2/css/all.min.css" integrity="" crossorigin="anonymous"/>
<link rel="stylesheet" href="/style_rcp-e600e6.css?t=1744906951"/>

<style>
    body {
        font-family: system-ui, -apple-system, "Segoe UI", Roboto, "Helvetica Neue", "Noto Sans", "Liberation Sans", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
    }
    .hidden {
        display:none;
    }
    
    #asdf {
        width: 0px;
        height: 0px;
    }
    
    #iframe_title {
        position: absolute;
        top: calc(50% + 45px);
        font-size: 1.1em;
        text-shadow: -1px -1px 0 #000, 1px -1px 0 #000, -1px 1px 0 #000, 1px 1px 0 #000;
        width: calc(100% - 30px);
        text-align: center;
        padding: 0px 15px;
    }
</style>
</head>
<body data-i="2085059_4x2"  >
    
    
        <!--<div id="pop_asdf" style="position:absolute; width:100%; height:100%; z-index:2147483650; pointer-events: none;"></div>-->
        
    <!--<div id="loading_overlay"></div>-->
    <div id="the_frame" style="background-image: none;">
                <div id="pl_but_background" style="background-image: url('//image.tmdb.org/t/p/w780/dg3OindVAGZBjlT3xYKqIAdukPL.jpg');">
            <div id="iframe_title">
                Black Mirror (2011) S04E02            </div>
            <i id="pl_but" class="fas fa-play"></i>
        </div>
            </div>
    
    <div id="hidden" style="display:hidden;"></div>
    
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js" integrity="sha256-1A78rJEdiWTzco6qdn3igTBv9VupN3Q1ozZNTR4WE/Y=" crossorigin="anonymous"></script>
    <script src="/base64.js?t="></script>
    
    <script src="/sbx.js?t=1751380596"></script>

    
    <script>
    
    /*
    $(document).ready(function(){
        
        
        if($.cookie('pop_asdf')){
            $("#pop_asdf").addClass("hidden");
        }
        
        $("#pop_asdf").click(function(){
            if(!$.cookie('pop_asdf')){
                if($.cookie('pop_asdf_tmp') >= 3){
                    var date = new Date();
                    date.setTime(date.getTime() + (10800 * 1000));
                    $.cookie('pop_asdf', 1, { expires: date });
                    $("#pop_asdf").addClass("hidden");
                }else{
                    var cookie_value = 1;
                    if($.cookie('pop_asdf_tmp'))
                        cookie_value = $.cookie('pop_asdf_tmp');
                        
                    cookie_value++;
                    
                    $.cookie('pop_asdf_tmp', cookie_value);
                    $(this).addClass("hidden");
                    setTimeout(function(){
                        $("#pop_asdf").removeClass("hidden");
                    }, 59000);
                }
            }
        });
    });
    */
    
        
    $("#pl_but_background , #pl_but").click(function(){
        loadIframe();
    });
    
        
    function loadIframe(data = 1){
        if(data == 1){
            $("#the_frame").removeAttr("style");
            $("#the_frame").html("");
            $('<iframe>', {
               id: 'player_iframe',
               src: '/prorcp/ZTUxNjZlOGE1MWIwYWNkNTdlZDFmYzU1OTY2MGQ0YTc6VDNWT1pFZHVNRGxxT0hrekswWkxlUzlvUnpSdGIweFJVMkUzYkhOTlFXTnhNM1pvWm5aTlYzY3ZVM1JXTjB0d0wyVjFLek5oTTNkVWRtWm9XRzFXWVdGUk5HbzRlRFowVld3elYzUm9WblJtUjNwdWNUWTNTSEJPVlZKdk5sVjJlVlI2ZVZkbFpESmlRV0ZYUVdSelpVZEpZbEZwWXpSWFVXeExTSEJxVVZGTGRrdGFkSFYxZFRaV2VrZDNNMngxSzNKaWVEZGxSbXg0TnpOSWRtVnpTM1JoYkc4ME1VZGFiVkI0UzA5M1l5OW9lRXdyYVc5U01HVlljREZLZWtremRsaEZhMjVZYmtSWlQzcFBlRzl5YjJOQ1VUVmlNbTVzVWtWSU9FaHNZek5VT1dSUGJFWlhZV1JYVVZSd01UYzNhMFZHYWt4SFp6VnpPVkZRT1VGT1NUaFJSRTFuTmxWdWFHcHRjRGx5YjFSMlNXaERaMFE1VEd4UVJHazRWMHRJUmxNME1sUXZWVU5DYW5oUlZFcHhhM00yVjNkUFRqSnpVbXBHVDAxWWVHUmtPV1EyUW5vMVJEWTRTMGhPYWsxcGVXaDNVRUpVTW5Wck5YZDRkSEZUWVhWbmVHa3lNbkZ5WVZGcVltMW1WbVJDYTAxemN6aHBiMlpGTW01V1N6aEpjbWhrWjFvd1RUUkROaXN2VFZCMVQyeHdWMjgxYkVaNlVuSlFjbE40WTFBeVlXbGFWMnBoYjNaUVlTOVVTbVJNUkM5RFNuZEZWSFZWTkhGRFkwWTBVbk5RYW1Kb1NWWXJibU5NY1V4QlJUbE1VbE4yWmk5NmFVMWlUalpuYzJFMlVUTlZPSGRTY0c5WmR6VkRTSEp3VEhwVWNsTlJXbTlhYzAxdFpraGtMMDF3YmxWdVZXdHBVSGhNSzNoVk1IUm9kVlZ0VVZRelFWZE9ZbGRIWVZFck5ERXhTVGxEZFRreGQzRjFkVTFKUkU0eU9XVlBVRmQ0TUhSU2Myc3hjSGxxVnlzMFMwZHJkMlp5V25KSVJrUkhWamRaWjJaWFNGUkZNWEpZUlVaR05EaGxla0ZCTTFSVkt5ODJkRmR2VkRoUlNuWkZWa2hhVkVkSlZtSkRUVnBzWW5wSFdXRnNkM0pCSzNkR2RVSnBkalpSVlhWWVNVZG1WR1o1WVZoaGFYWkpTRmRvV2xWclNFSTBVbTl6WWtabk4xbGFRazlTVmt0U2Rqa3dRMFZSYTFkU1ZsUXlTVFYyVVVrNVZqSlpOSGxJU2poM1dISTFkazVXUVZoVmMxcE1TV2MxWTNsbGQwaG9PREZ3WkhsbFowMURjVmxOZEdJMVJqUnVTR0pMZGsxSlZrMXlaMGN5UmxWTlNFMHhaWEpuWjFWNVRUWkpTekk1V1haRVNHMU1lbVJyWWxFd1NWTmlZemRVVTNKRFRWWjNWWGhTZVVoTmMyeGpXbGh3T0hFeVpFRXlZaXRQT1c4d1ZYZEZLMjFOVDI5NFdVWjFUVGsxT1c4elQzSmhXSHBhU1hwV1RUSllOVEpwVTJ4WU5HVlVhVFJVZFhGS2VrMWpTbmg2ZW1Sdk1scHZiRGh6WW1aTWQyVm9RMmxQZVdFeVoyOXViMEV5YW1GbmFERmhZbEprYnpaelZUSkJTM295Wm1Kd1dtUXhaM1JOYUU5SGRtSktRakZqVDBSU1NYWjViM1JvU1VKcFZrOWxUM0ZWTHpJMksydEVLMk5qWmt0Q2IydFFkVVZzTnpCQlduaExXVGxuUm5sTFFtWnBXSGhvWld0MFRqazJRbFJZYTJSQ09FRmpRMlJyY21oc2MyNW5TMHR5VUZaVVRrWlhUbmhEZEVZeU9EWnhSMGhrWVdaa1JTOVhOM0ZZU1RSWGNrZ3ZWa2R1VkhGUlFrMVZObUV4WW1Wb1MwaEVhRXhXUWpCNlptSk1RVXN6U1M5WldISklTalZhYW1wRVMwcEJPVmRKUlU1R1UxTnVOR1pQV1dWNVdrY3pUMFZDY25WMmMyRlhWREZzY0Vjd2RIVnlTMUJ4TWs4dk5UVnVaalJ2TjJScVR6UjViVU01ZWxRNVMyWnViVmt6YUZwVUswUTBSV1JvTkRGalNub3dNRVphTlZSSlpUaDZkVlJ3UzNOWlNrSmlWbTlXY2pVd1ltcDNVRzF4UVRCdmIzSklWMFJ3UzAxbU5IYzlQUT09',
               frameborder: 0,
               scrolling: 'no',
               allowfullscreen: 'yes',
               allow: "autoplay",
               style: 'height: 100%; width: 100%;'
            }).appendTo('#the_frame');
            $("#player_iframe").on("load", function () {
                $("#the_frame").attr("style","background-image: none;");
            });
        }
    }
    
    // pm redirector
    window.addEventListener('message', message => {
        if (message.source == window) {
            return; // Skip message in this event listener
        }
        
        var the_iframe = document.getElementById('player_iframe');
        
        if(message.source == window.parent){
            the_iframe.contentWindow.postMessage(message.data,'*');
        }else{
            window.parent.postMessage(message.data , '*');
        }
        
        
        
    });
        
        
        
    </script>
    
    
    
    
    
    
   
</body>
</html>
