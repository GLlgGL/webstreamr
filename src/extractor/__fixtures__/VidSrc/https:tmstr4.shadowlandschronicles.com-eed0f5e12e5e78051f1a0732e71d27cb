#EXTM3U
#EXT-X-INDEPENDENT-SEGMENTS
#EXT-X-STREAM-INF:BANDWIDTH=833322,CODECS="mp4a.40.2,avc1.42c01e",RESOLUTION=640x358,FRAME-RATE=24,VIDEO-RANGE=SDR,CLOSED-CAPTIONS=NONE
/pl/H4sIAAAAAAAAAw3Oy3KCMBhA4VcKBqh2p0UumRJKSP4Iu1ywkYDDUKuVp6_bs_jmWBVtg.3ZvMUoDqLAxOps.jDCsVHIbPrdu5FuqQQ5UGS_dELbZh3vegVfivna5I7wqcv6nJEuYKzGaUmn71Ajv1qAR4NSX51YLjOWdN6qXsKxOxWou76cqbzzy.63Pu2fkJCjSI64kuViE3fRGZl1OjvmU8cxYOPtR7u5PYDbVg0uEANU0rPLJyofYmBcJPu_Buif5OBoOi8tBsowKRhmaSPRUw2v7n9WO76uspno3Ec8LxYeHA5dAKGU3RM2YaibGzbZ7MqxCJVkmZJutdc6qhGROgNZrqnQIxsttgVMdBEc7f4BOJFmy0EBAAA-/f2a0a2e3ac935bed168e40c6c28105d7/index.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=2728709,CODECS="mp4a.40.2,avc1.64001f",RESOLUTION=1280x714,FRAME-RATE=24,VIDEO-RANGE=SDR,CLOSED-CAPTIONS=NONE
/pl/H4sIAAAAAAAAAw3Oy3KCMBhA4VcKBqh2p0UumRJKSP4Iu1ywkYDDUKuVp6_bs_jmWBVtg.3ZvMUoDqLAxOps.jDCsVHIbPrdu5FuqQQ5UGS_dELbZh3vegVfivna5I7wqcv6nJEuYKzGaUmn71Ajv1qAR4NSX51YLjOWdN6qXsKxOxWou76cqbzzy.63Pu2fkJCjSI64kuViE3fRGZl1OjvmU8cxYOPtR7u5PYDbVg0uEANU0rPLJyofYmBcJPu_Buif5OBoOi8tBsowKRhmaSPRUw2v7n9WO76uspno3Ec8LxYeHA5dAKGU3RM2YaibGzbZ7MqxCJVkmZJutdc6qhGROgNZrqnQIxsttgVMdBEc7f4BOJFmy0EBAAA-/ea18a1816b821e1ba4ee5d4a76b0b7cf/index.m3u8