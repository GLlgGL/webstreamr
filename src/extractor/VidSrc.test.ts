import fetchMock from 'fetch-mock';
import winston from 'winston';
import { createTestContext } from '../test';
import { Fetcher, FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { VidSrc } from './VidSrc';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new VidSrc(new FetcherMock(`${__dirname}/__fixtures__/VidSrc`), ['xyz'])]);

const ctx = createTestContext();

describe('VidSrc', () => {
  beforeAll(() => {
    fetchMock.mockGlobal();
  });

  afterAll(() => {
    fetchMock.unmockGlobal();
  });

  afterEach(() => {
    fetchMock.clearHistory();
  });

  test('Full Metal Jacket', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://vidsrc.xyz/embed/movie/tt0093058'))).toMatchSnapshot();
  });

  test('Black Mirror', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://vidsrc.xyz/embed/tv/tt2085059/4/2'))).toMatchSnapshot();
  });

  test('not found', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://vidsrc.xyz/embed/movie/tt35628853'))).toMatchSnapshot();
  });

  test('rate limit issues are retried and fail if no tlds are left', async () => {
    fetchMock
      .getOnce('https://vidsrc.xyz/embed/movie/tt33043892/1/1', 429)
      .getOnce('https://vidsrc.net/embed/movie/tt33043892/1/1', 429);

    const fetcher = new Fetcher(winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] }));
    const vidSrc = new VidSrc(fetcher, ['net', 'xyz']);

    expect(await vidSrc.extract(ctx, new URL('https://vidsrc.xyz/embed/movie/tt33043892/1/1'), {})).toMatchSnapshot();
  });

  test('blocking issues are retried and fail if no tlds are left', async () => {
    fetchMock
      .getOnce('https://vidsrc.xyz/embed/movie/tt33043892/1/2', 403)
      .getOnce('https://vidsrc.net/embed/movie/tt33043892/1/2', 403);

    const fetcher = new Fetcher(winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] }));
    const vidSrc = new VidSrc(fetcher, ['net', 'xyz']);

    expect(await vidSrc.extract(ctx, new URL('https://vidsrc.xyz/embed/movie/tt33043892/1/2'), {})).toMatchSnapshot();
  });
});
