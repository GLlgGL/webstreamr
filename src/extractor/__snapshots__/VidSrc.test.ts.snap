// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`VidSrc Black Mirror 1`] = `
[
  {
    "format": "hls",
    "label": "VidSrc (CloudStream Pro)",
    "meta": {
      "countryCodes": [],
      "height": 1040,
      "title": "Black Mirror (2011) S04E02",
    },
    "sourceId": "vidsrc_CloudStream-Pro_",
    "ttl": 10800000,
    "url": "https://tmstr5.shadowlandschronicles.com/pl/H4sIAAAAAAAAAxXK226DIAAA0F8CrGvd27oCq1UcyP1NQdtUaozptsyvX3aez9DBQxgjgBmAPeqK_Yh2eTeG4tAPMCDwysHNaeDbevKVMM.XgMWps2XtTWGkLR8M.feQ2KroeeXG_db4vCq82.T8lkerZ0OTlLLsWuA2lnwT1FK7e6r0B5nEHL_4pNAgme2sJj28_kT0f0kuya0Sj6W6wONRnkrpAHP9vZwEzRE3T8JOohmwtwMlTimNeiw2BQWRsM4CKZ3PFuUyT1usLy6LWb3F1s3p0hPRSpqqxibcPpaG3.Mq5qS4vn4bqm.NTp8KFzOfmBykWP4AirgCPSEBAAA-/master.m3u8",
  },
]
`;

exports[`VidSrc Full Metal Jacket 1`] = `
[
  {
    "format": "hls",
    "label": "VidSrc (CloudStream Pro)",
    "meta": {
      "countryCodes": [],
      "height": 714,
      "title": "Full Metal Jacket (1987)",
    },
    "sourceId": "vidsrc_CloudStream-Pro_",
    "ttl": 10800000,
    "url": "https://tmstr4.shadowlandschronicles.com/pl/H4sIAAAAAAAAAwXB7XqCIBgA0FtCsa_97IMcFSa.gPBPwc1H0Fm5Wl79zllVS4eRdZv10uJkuarrJF7Yr8at1lWyRvZDC7cFP3IT2YU4PHAmiGKB59C1Z41bbUQ7czDQ.AgpMv6csHtqNTFGzMMGessV_bOSvRkadab4riBhWx0nxuPxDtFnXMSjqAOdM5AtK3lyjjxiyD.LYAaRbqvC8_MF_L0Bsq8GeuGHxUlLU9rSZapvkSTyUfsNkcEJi8eb9jRpRPJkMTO6H4mV33cnaCqP_LeIpw46iVlHrxDp2c6G5ILtbeoC9I42itxy_3plqev50N5FLDHfTS94T1AQKXUZrmwOA1dtVIFFOaDNP7JZ1PhBAQAA/master.m3u8",
  },
]
`;

exports[`VidSrc not found 1`] = `[]`;

exports[`VidSrc rate limit issues are retried and fail if no tlds are left 1`] = `
[
  {
    "error": [Error],
    "format": "unknown",
    "isExternal": true,
    "label": "VidSrc",
    "sourceId": "vidsrc",
    "url": "https://vidsrc.xyz/embed/movie/tt33043892/1/1",
  },
]
`;
